server/src/main/kotlin/com/trib3/server/graphql/GraphQLWebSocket.kt

```diff
@@ -92,16 +182,30 @@ open class GraphQLWebSocket(
      */
     private fun handleConnectionInit(message: OperationMessage<*>) {
         sendMessage(OperationType.GQL_CONNECTION_ACK, message.id)
+        keepAliveLock.withLock {
+            if (this.keepAliveInterval == null) {
+                sendMessage(OperationType.GQL_CONNECTION_KEEP_ALIVE, message.id)
+                val keepAliveCallbacks = KeepAliveCallbacks(this, message.id)
+                this.keepAliveInterval = Observable.interval(15, TimeUnit.SECONDS, Schedulers.io())
+                    .subscribe(keepAliveCallbacks::onInterval, keepAliveCallbacks::onError)
+            }
+        }
     }
 
     /**
      * Stop any running query and terminate the connection
      */
-    private fun handleConnectionTermination() {
+    private fun handleConnectionTermination(statusCode: Int, reason: String?) {
         runningQueryLock.withLock {
             runningQuerySubscriber?.unsubscribe()
             runningQuerySubscriber = null
-            session.close(StatusCode.NORMAL, "Termination Requested")
+            keepAliveLock.withLock {
```

Comment:
any danger of keepalivelock waiting on runningquerylock?

Reply:
both locks are held for relatively small amounts of time, and the ordering is never reversed, so i think it's safe, but not ever as confident as I'd like to be given it is multithreaded locking code.... 

kotlin prefers explicit lock.withLock {} statements over the java synchronized blocks based on what i've found (it actually might not even support synchronized!)....